File: benchmarks/pandas/bench_dropna_advanced.py
"""
Benchmark: DataFrame.dropna with advanced options (thresh, subset, axis=1).
Outputs JSON: {"function": "dropna_advanced", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import numpy as np
import pandas as pd

SIZE = 10_000
WARMUP = 5
ITERATIONS = 30

# DataFrame with scattered null values
rng = np.random.default_rng(42)
df = pd.DataFrame({
    "a": [None if i % 4 == 0 else i * 0.1 for i in range(SIZE)],
    "b": [None if i % 6 == 0 else i * 2.0 for i in range(SIZE)],
    "c": [None if i % 8 == 0 else i % 100 for i in range(SIZE)],
    "d": [None if i % 3 == 0 else f"val_{i % 20}" for i in range(SIZE)],
})

for _ in range(WARMUP):
    df.dropna(thresh=3)
    df.dropna(subset=["a", "b"])
    df.dropna(axis=1)

start = time.perf_counter()
for _ in range(ITERATIONS):
    df.dropna(thresh=3)
    df.dropna(subset=["a", "b"])
    df.dropna(axis=1)
total = (time.perf_counter() - start) * 1000

print(json.dumps({"function": "dropna_advanced", "mean_ms": total / ITERATIONS, "iterations": ITERATIONS, "total_ms": total}))

File: benchmarks/tsb/bench_dropna_advanced.ts
/**
 * Benchmark: dropnaDataFrame with advanced options (thresh, subset, axis=1).
 * Outputs JSON: {"function": "dropna_advanced", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { DataFrame, dropnaDataFrame } from "../../src/index.ts";

const SIZE = 10_000;
const WARMUP = 5;
const ITERATIONS = 30;

// DataFrame with scattered null values
const df = DataFrame.fromColumns({
  a: Array.from({ length: SIZE }, (_, i) => (i % 4 === 0 ? null : i * 0.1)),
  b: Array.from({ length: SIZE }, (_, i) => (i % 6 === 0 ? null : i * 2.0)),
  c: Array.from({ length: SIZE }, (_, i) => (i % 8 === 0 ? null : i % 100)),
  d: Array.from({ length: SIZE }, (_, i) => (i % 3 === 0 ? null : `val_${i % 20}`)),
});

for (let i = 0; i < WARMUP; i++) {
  dropnaDataFrame(df, { thresh: 3 });
  dropnaDataFrame(df, { subset: ["a", "b"] });
  dropnaDataFrame(df, { axis: 1 });
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  dropnaDataFrame(df, { thresh: 3 });
  dropnaDataFrame(df, { subset: ["a", "b"] });
  dropnaDataFrame(df, { axis: 1 });
}
const total = performance.now() - start;

console.log(JSON.stringify({ function: "dropna_advanced", mean_ms: total / ITERATIONS, iterations: ITERATIONS, total_ms: total }));
